### PR TITLE
Debug Test update

### PR DESCRIPTION
**Pull request by quasiben**
supersedes https://github.com/rapidsai/ucx-py/pull/453

PR changes the debug test to mimic a "broadcast" in dask.  send.py acts as the "scheduler" and recv.py acts as a "worker".  Additionally, I've added an example bash script which deploys a multi-worker setup

## Process 1

```
export UCX_LOG_LEVEL=TRACE
export UCXPY_LOG_LEVEL=DEBUG
export UCX_MEMTYPE_CACHE=n
export UCX_TLS=tcp,rdmacm,cuda_copy,rc
export UCX_SOCKADDR_TLS_PRIORITY=rdmacm

UCX_NET_DEVICES=mlx5_0:1 CUDA_VISIBLE_DEVICES=0 python send.py 2>&1 | tee /tmp/send-log.txt &
```

## Process 2
```
bash multi-node-ib-test.sh
```

PR is still WIP do not merge

### CODE REVIEW

**Comment by quasiben** [comment]
The multinode test fails with 4 DGX1s with the error:

```
[1586794449.499549] [prm-dgx-04:48143:1]   sockcm_iface.c:221  UCX  DEBUG sockcm_iface 0x556c6a8f9af0: accepted connection from 10.33.12.16:59218 at fd 1118 Resource temporarily unavailable
[prm-dgx-04:48143:1:48161] sockcm_iface.c:257  Fatal: sockcm_listener: unable to create handler for new connection
==== backtrace ====
size of the message:  23
ITER:  26
DEBUG [Send #156] ep: 0x7f7dd09c4a68, tag: 0x15054f8c605dd8e6, nbytes: 8
    0  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(ucs_fatal_error_message+0x51) [0x7f7e362d2171]
    1  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(ucs_fatal_error_format+0xf5) [0x7f7e362d2325]
    2  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libuct.so.0(+0x1fddf) [0x7f7e36331ddf]
    3  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(+0x1021a) [0x7f7e362c321a]
    4  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(+0x103a8) [0x7f7e362c33a8]
    5  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(ucs_async_dispatch_handlers+0x30) [0x7f7e362c3550]
    6  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(+0x14008) [0x7f7e362c7008]
    7  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(ucs_event_set_wait+0x105) [0x7f7e362da495]
    8  /gpfs/fs1/bzaitlen/miniconda3/envs/20200330/lib/libucs.so.0(+0x141c0) [0x7f7e362c71c0]
    9  /lib/x86_64-linux-gnu/libpthread.so.0(+0x76db) [0x7f7edb7046db]
   10  /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f7edb42d88f]
===================
Aborted (core dumped)
```

**Comment by quasiben** [comment]
We I reduce the iterations to 3 and the program "successfully" finishes, on exit, things get complicated for the scheduler:


```
667ab8: created next_ep 0x55659a5dbdc0 to <no debug data> using rc/mlx5_0:1
[1586812594.776325] [prm-dgx-12:6228 :0]          ud_ep.c:208  UCX  DEBUG created ep ep=0x55659a7bf470 iface=0x55657887ae30 id=94
[1586812594.776328] [prm-dgx-12:6228 :0]          ud_ep.c:294  UCX  DEBUG mlx5_0:1 lid 107 qpn 0x537 epid 94 ep 0x55659a7bf470 connected to IFACE lid 134 fe80:: qpn 0x23087
[1586812594.776333] [prm-dgx-12:6228 :0]      wireup_ep.c:298  UCX  DEBUG ep 0x7f9ea4667ab8: wireup_ep 0x5565993e30c0 created aux_ep 0x55659a7bf470 to <no debug data> using ud/mlx5_0:1
[1586812594.776335] [prm-dgx-12:6228 :0]         wireup.c:982  UCX  DEBUG ep 0x7f9ea4667ab8: send wireup request (flags=0x400080)
[1586812594.777190] [prm-dgx-12:6228 :0]       rc_iface.c:808  UCX  DEBUG connected rc qp 0x59f on mlx5_0:1 to lid 128(+0) sl 0 remote_qp 0x29e35 mtu 4096 timer 18x7 rnr 13x7 rd_atom 4
[1586812594.777196] [prm-dgx-12:6228 :0]      wireup_ep.c:376  UCX  DEBUG ep 0x7f9ea4667a70: destroy wireup ep 0x556599483600
[1586812594.777198] [prm-dgx-12:6228 :0]          ud_ep.c:1287 UCX  DEBUG ep 0x55659a7bf360: disconnect
[1586812594.777872] [prm-dgx-12:6228 :0]       rc_iface.c:808  UCX  DEBUG connected rc qp 0x5a0 on mlx5_0:1 to lid 134(+0) sl 0 remote_qp 0x2308d mtu 4096 timer 18x7 rnr 13x7 rd_atom 4
[1586812594.777879] [prm-dgx-12:6228 :0]      wireup_ep.c:376  UCX  DEBUG ep 0x7f9ea4667ab8: destroy wireup ep 0x5565993e30c0
[1586812594.777894] [prm-dgx-12:6228 :0]          ud_ep.c:1287 UCX  DEBUG ep 0x55659a7bf470: disconnect
[1586812594.791213] [prm-dgx-12:6228 :0]         ucp_ep.c:131  UCX  DEBUG created ep 0x7f9ea4667b00 to <no debug data> listener
[1586812594.791225] [prm-dgx-12:6228 :0]         wireup.c:796  UCX  DEBUG ep 0x7f9ea4667b00: am_lane 0 wireup_lane 0 reachable_mds 0x2
[1586812594.791230] [prm-dgx-12:6228 :0]         wireup.c:802  UCX  DEBUG ep 0x7f9ea4667b00: lane[0]:  1:rc/mlx5_0:1 md[1]           -> addr[1].md[1]/ib       rma_bw#0 am am_bw#0 wireup
[1586812594.791477] [prm-dgx-12:6228 :0]       ib_iface.c:598  UCX  DEBUG iface=0x556578b038d0: created RC QP 0x5a1 on mlx5_0:1 TX wr:409 sge:8 inl:124 RX wr:0 sge:0 inl 64
[1586812594.791888] [prm-dgx-12:6228 :0]      wireup_ep.c:437  UCX  DEBUG ep 0x7f9ea4667b00: created next_ep 0x55659a5dbe20 to <no debug data> using rc/mlx5_0:1
[1586812594.791901] [prm-dgx-12:6228 :0]          ud_ep.c:208  UCX  DEBUG created ep ep=0x55659a7bf580 iface=0x55657887ae30 id=95
[1586812594.791910] [prm-dgx-12:6228 :0]          ud_ep.c:294  UCX  DEBUG mlx5_0:1 lid 107 qpn 0x537 epid 95 ep 0x55659a7bf580 connected to IFACE lid 116 fe80:: qpn 0x28302
[1586812594.791955] [prm-dgx-12:6228 :0]      wireup_ep.c:298  UCX  DEBUG ep 0x7f9ea4667b00: wireup_ep 0x5565993e30c0 created aux_ep 0x55659a7bf580 to <no debug data> using ud/mlx5_0:1
[1586812594.791960] [prm-dgx-12:6228 :0]         wireup.c:982  UCX  DEBUG ep 0x7f9ea4667b00: send wireup request (flags=0x400080)
[1586812594.794775] [prm-dgx-12:6228 :0]       rc_iface.c:808  UCX  DEBUG connected rc qp 0x5a1 on mlx5_0:1 to lid 116(+0) sl 0 remote_qp 0x2830a mtu 4096 timer 18x7 rnr 13x7 rd_atom 4
[1586812594.794782] [prm-dgx-12:6228 :0]      wireup_ep.c:376  UCX  DEBUG ep 0x7f9ea4667b00: destroy wireup ep 0x5565993e30c0
[1586812594.794811] [prm-dgx-12:6228 :0]          ud_ep.c:1287 UCX  DEBUG ep 0x55659a7bf580: disconnect
[1586812608.771538] [prm-dgx-12:6228 :0]      uct_iface.c:326  UCX  DEBUG set ep 0x55659a7b72d0 to failed state
[1586812608.772097] [prm-dgx-12:6228 :0]          rc_ep.c:127  UCX  DEBUG destroy rc ep 0x55659a7b72d0
[1586812608.772101] [prm-dgx-12:6228 :0]     ucp_worker.c:569  UCX  DEBUG worker 0x5565731373d0: error handler called for uct_ep 0x55659a7b72d0: Endpoint timeout
[prm-dgx-12:6228 :0:6228] rc_verbs_iface.c:69   send completion with error: transport retry counter exceeded qpn 0x587 wrid 0x1 vendor_err 0x81
```

File: tests/debug-tests/send.py [hunk 1]
```diff
@@ -32,31 +31,28 @@ def server(env, port, func):
...
 
-            msg = np.empty(msg_size[0], dtype=np.uint8)
-            await ep.recv(msg)
-            recv_msg = msg.tobytes()
-            assert recv_msg == close_msg
-            print("Shutting Down Server...")
-            await ep.close()
-            lf.close()
+                print("CONFIRM RECEIPT")
+                await ep.close()
```
Comment (madsbk) [inline tests/debug-tests/send.py:52]: I only see the bug when when we try to send using a closed endpoint. Exiting the loop after closing the endpoint avoids the bug:
```suggestion
                await ep.close()
                break
```
However, exiting the loop via. a `ucp.exceptions.UCXCloseError` should be perfectly legal.

Comment (quasiben) [reply tests/debug-tests/send.py:52]: Hmm, in this test, I wanted to keep the sender alive forever.   Breaking here will close the sender, right ?

Comment (madsbk) [reply tests/debug-tests/send.py:52]: it breaks in the next `send()` call anyways because `ep` is closed.

**Comment by jakirkham** [comment]
There's an `isort` failure that points to `test_send_recv_two_workers.py`. Not sure how important that is to fix yet (given this is WIP).

**Comment by quasiben** [comment]
I am trying to use this PR to debug some of the endpoint failures.  Not trying to merge as of yet

**Comment by pentschev** [comment]
I think this is quite useful, can we get it merged @quasiben ? I can also take this over if you prefer.

**Comment by quasiben** [comment]
Yes, I can work on this next week and get it merged in

**Comment by pentschev** [comment]
Thanks @quasiben !

**Comment by quasiben** [comment]
@madsbk if you have time can you review ?

**Comment by quasiben** [comment]
rerun tests

File: tests/test_send_recv_two_workers.py [hunk 2]
```diff
@@ -103,12 +103,12 @@ async def read():
...
-        msg = f"RX BEFORE SEND: {before_rx} -- RX AFTER SEND: {rx} \
-               -- TOTAL DATA: {num_bytes}"
-        print(msg)
-        assert rx > before_rx
+    # if num_bytes > 90000:
+    #     rx, tx = total_nvlink_transfer()
+    #     msg = f"RX BEFORE SEND: {before_rx} -- RX AFTER SEND: {rx} \
+    #            -- TOTAL DATA: {num_bytes}"
+    #     print(msg)
+    #     assert rx > before_rx
```
Comment (madsbk) [inline tests/test_send_recv_two_workers.py:111]: Should we remove this completely?

Comment (quasiben) [reply tests/test_send_recv_two_workers.py:111]: yes!  I'll do that now

**Comment by quasiben** [comment]
Thanks @mads .  Merging in

**Pull request merged**